ocaml/xapi-guard/lib/server_interface.ml

```diff
@@ -99,109 +100,65 @@ let serve_forever_lwt_callback rpc_fn path _ req body =
-let split_char = ' '
-
-let join_string = String.make 1 split_char
-
-let deserialize t =
-  match String.split_on_char split_char t with
-  | [permall] ->
-      (* backwards compat with reading tpm2-00.permall *)
-      {permall; savestate= ""; volatilestate= ""}
-  | [permall; savestate; volatilestate] ->
-      {permall; savestate; volatilestate}
-  | splits ->
-      Fmt.failwith "Invalid state: too many splits %d" (List.length splits)
-
-let serialize t =
-  (* it is assumed that swtpm has already base64 encoded this *)
-  String.concat join_string [t.permall; t.savestate; t.volatilestate]
-
-let lookup_key key t =
-  match key with
-  | "/tpm2-00.permall" ->
-      t.permall
-  | "/tpm2-00.savestate" ->
-      t.savestate
-  | "/tpm2-00.volatilestate" ->
-      t.volatilestate
-  | s ->
-      Fmt.invalid_arg "Unknown TPM state key: %s" s
-
-let update_key key state t =
-  if String.contains state split_char then
-    Fmt.invalid_arg
-      "State to be stored (%d bytes) contained forbidden separator: %c"
-      (String.length state) split_char ;
-  match key with
-  | "/tpm2-00.permall" ->
-      {t with permall= state}
-  | "/tpm2-00.savestate" ->
-      {t with savestate= state}
-  | "/tpm2-00.volatilestate" ->
-      {t with volatilestate= state}
-  | s ->
-      Fmt.invalid_arg "Unknown TPM state key: %s" s
-
-let empty = ""
-
-let serve_forever_lwt_callback_vtpm ~cache mutex vm_uuid _ req body =
-  let get_vtpm_ref () =
-    let* vm =
-      with_xapi ~cache @@ Xen_api_lwt_unix.VM.get_by_uuid ~uuid:vm_uuid
-    in
-    let* vTPMs = with_xapi ~cache @@ Xen_api_lwt_unix.VM.get_VTPMs ~self:vm in
-    match vTPMs with
-    | [] ->
-        D.warn
-          "%s: received a request from a VM that has no VTPM associated, \
-           ignoring request"
-          __FUNCTION__ ;
-        let msg =
-          Printf.sprintf "No VTPM associated with VM %s, nothing to do" vm_uuid
-        in
-        raise (Failure msg)
-    | self :: _ ->
-        let* uuid = with_xapi ~cache @@ Xen_api_lwt_unix.VTPM.get_uuid ~self in
-        with_xapi ~cache @@ VTPM.get_by_uuid ~uuid
+let with_xapi_vtpm ~cache vm_uuid =
+  let vm_uuid_str = Uuidm.to_string vm_uuid in
+  let* vm =
+    with_xapi ~cache @@ Xen_api_lwt_unix.VM.get_by_uuid ~uuid:vm_uuid_str
   in
+  let* vTPMs = with_xapi ~cache @@ Xen_api_lwt_unix.VM.get_VTPMs ~self:vm in
+  match vTPMs with
+  | [] ->
+      D.warn
+        "%s: received a request from a VM that has no VTPM associated, \
+         ignoring request"
+        __FUNCTION__ ;
+      let msg =
+        Printf.sprintf "No VTPM associated with VM %s, nothing to do"
+          vm_uuid_str
+      in
+      raise (Failure msg)
+  | self :: _ ->
+      Lwt.return self
+
+let push_vtpm ~cache (vm_uuid, _timestamp, key) contents =
+  let* self = with_xapi_vtpm ~cache vm_uuid in
+  let* old_contents = with_xapi ~cache @@ VTPM.get_contents ~self in
+  let contents =
+    Tpm.(old_contents |> deserialize |> update key contents |> serialize)
```

Comment:
This does a read-modify-write cycle on each update for the 3 separate TPM entrances. I think it was done this way for backwards compatibility (and now I wish I hadn't done it this way), but is bad for the cache, because this means that every write to any of the 3 TPM states requires to read the *latest* value of the other 2 states.

I haven't checked the code, but according to the design docs I think this could lead to the following bug and dataloss (and it should've been possible to catch this bug with a quickcheck-stm test):
* start out with vTPM states X=X0,Y=Y0,Z=Z0
* xapi down
* write to the vTPM state X = X1, state is now X=X1,Y=Y0,Z=Z0
* write to the vTPM state Y = Y1, state is now X=X0,Y=Y1,Z=Z0
* write to the vTPM state Z = Z1, state is now X=X0,Y=Y0,Z=Z1
* reboot VM
* xapi up
* state has now been lost, we have Z=Z1, but we've lost X1 and Y1

it is not immediately obvious how to trigger this through the TPM API (the 3 states serve different purposes), but to avoid this bug I think we need to:
* store the 3 states in 3 separate caches so that we don't require read-modify-write cycles
* fix the cache, so that when you read back something after a write you get the latest value (this is how a cache is supposed to behave).

Reply:
> store the 3 states in 3 separate caches so that we don't require read-modify-write cycles

The disk cache stores the updates, not the whole contents of the TPM. This is because we don't want writes to disk to depend on xapi.